### PR TITLE
Apply energy loss cut *after* slowing down rather than before

### PR DESCRIPTION
**Pull request by sethrj**
This prevents ultra-low-energy charged particles from sticking around after energy loss. In pathological situations (range slightly longer than boundary step) we can still hypothetically get such particles, but they should be vanishingly unlikely.

Should make #616 and #619 much more unlikely to be triggered.

### CODE REVIEW

File: src/celeritas/global/alongstep/AlongStep.hh [hunk 1]
```diff
@@ -108,17 +109,83 @@ inline CELER_FUNCTION void along_step(MH&& msc,
...
+                   <= value_as<Energy>(
+                       track.make_physics_view().scalars().eloss_calc_limit))
+        {
+            // The energy after slowing down is less than the hard tracking
+            // cutoff, so stop the particle rather than leaving it a tiny
+            // amount of energy.
+            // To avoid stopping particles on boundaries (which is an
+            // artificial bias), we avoid this for geometry-limited steps.
+            deposited = particle.energy();
+        }
```
Comment (whokion) [inline src/celeritas/global/alongstep/AlongStep.hh:129]: I find that this does not help for stopping the very lower energy e+/e- after along step actions, range/msc-limiter/boundary/msc-scatter/eloss (even though range is treated separately before along_step in celeritas).  So, let's apply `eloss_calc_limit` before `eloss.calc_eloss` (i.e., calculate the eloss above this cut, otherwise, eloss = particle energy as before). The rest looks okay to me and can be merged after @amandalund's review and approval.  As a follow up, let's review the whole e+/e- workflow and discuss how to properly stop very low energy e-/e+ below a certain energy threshold (a.k.a tracking cut) if necessary.  I can prepare a couple of slides for a follow up discussion at the next standup meeting.

**Review by whokion — CHANGES_REQUESTED** [review]

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: I'm confused. Applying before `eloss.calc_eloss` is what already is in the code, and it allows tracks that start *above* the cut to end up *below* the cut after slowing down, leading to low-energy tracks. Isn't this change (2507103c8b0c881613b4437cb2f22c33c9b9f9bd) what we discussed?

Comment (whokion) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: Yes, my suggestion is reverting the change (i.e., back to the original).  I thought that the source of the very low energy particle is from the case when the calculated eloss is very close to the particle energy (above the eloss_calc_limit cut) and the particle with a small remaining energy is queued to the next step, but it seems that there is something else - so, I propose not to impose the additional relaxation on the eloss condition until a real source of missing logic in the work flow is understood.

Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: OK, maybe the confusion is about when and where propagation versus energy loss happens?

>  I thought that the source of the very low energy particle is from the case when the calculated eloss is very close to the particle energy (above the eloss_calc_limit cut) and the particle with a small remaining energy is queued to the next step

That is what I believe is happening. In the current implementation, a particle that ends a step with very low energy will undergo propagation (on the next step) before it hits the `eloss_limit` criteria.

Comment (whokion) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: I believe that the current order of along step actions is right (i.e, the energy loss comes after the propagation and msc), but the status of track with the very small energy particle is not handled properly in the down stream (i.e., not either killed or stopped after depositing all its energy, so it still returns to the next along step even with this proposed change), which need to be understood.  Sorry for the confusion.

Comment (amandalund) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: Stopped positrons can reach the next step, but they will exit the along-step kernel early and not be propagated/undergo eloss etc.

I do see a very significant reduction in the number of very low energy (but not stopped) charged particles with this change (e.g. >99% fewer particles with energies below 1 eV).

File: src/celeritas/global/alongstep/detail/FluctELoss.hh [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using Energy = EnergyLossHelper::Energy;
+    using ParamsRef = NativeCRef<FluctuationData>;
+    //!@}
+
+  public:
+    // Construct with flucutation data
```
Comment (amandalund) [inline src/celeritas/global/alongstep/detail/FluctELoss.hh:34]:
```suggestion
    // Construct with fluctuation data
```

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (whokion) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: The stopped particle with `zero step length` exits at the beginning, which is good as it may be equivalently to the `zero energy` (stopped) particle (are these two are always exchangeable in celeritas?).

Okay, good to know that the cut rejects more than 99% of particle below 1eV.  Do you also know, what is the distribution of delta(energy-eloss) for primary particle energy below 1keV (eloss_calc_limit) and above 1keV.  The real question is why the new cut does not reject all particles below 1eV?  Who are those remaining low energy particles and where are they from?

File: src/celeritas/global/alongstep/AlongStep.hh [hunk 3]
```diff
@@ -108,17 +109,83 @@ inline CELER_FUNCTION void along_step(MH&& msc,
...
+            sim.status(TrackStatus::killed);
+        }
+        else
+        {
+            // Particle slowed down to zero: force a discrete interaction
+            local.step_limit.action = phys.scalars().discrete_action();
+        }
+    }
 
+    if (sim.status() != TrackStatus::killed)
```
Comment (amandalund) [inline src/celeritas/global/alongstep/AlongStep.hh:179]: Why didn't we need this check before?

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: The cut only applies to tracks when:
- undergoing energy loss (i.e. not after a discrete interaction, nor to gammas)
- aren't in the process of hitting a boundary (so that we don't stop particles on boundaries)

So we're more likely to have small-energy particles starting their first step from the boundary, but they're likely to die after the next propagation.

**Review by amandalund — APPROVED** [review]
Changes look good to me @sethrj (and nice to not have to maintain the duplicate code in the eloss appliers)

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (amandalund) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: > The stopped particle with zero step length exits at the beginning, which is good as it may be equivalently to the zero energy (stopped) particle (are these two are always exchangeable in celeritas?).

Not sure about always, but yes at the beginning of along-step they are equivalent.

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 3 above)
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:179]: I don't think we *need* it actually: it's just wasted computations? Unless this happens to stop the `mfp > 0` condition from being hit. I'm fine with leaving the conditional or removing it.

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: Looking at what Geant4 does: to be equivalent we would also need to apply the tracking cut at the boundaries, and after discrete interactions. I think the latter is reasonable but we shouldn't do the former (to avoid biasing gamma production to happen on geometry edges, which also creates problems with particle tracking in ORANGE). I think we end up doing a half-step less work than Geant4 if we apply the cut as this PR implements.

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 3 above)
Comment (amandalund) [reply src/celeritas/global/alongstep/AlongStep.hh:179]: Makes sense, we can keep it -- I just didn't know if it was something we were missing before or if there was some other change made it necessary.

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 1 above)
Comment (amandalund) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: > Looking at what Geant4 does: to be equivalent we would also need to apply the tracking cut at the boundaries, and after discrete interactions

Where are you seeing this? It looks like if it's an energy loss processes and the particle's energy is below this cutoff the discrete interaction is skipped, but I don't see where the cut is applied after an interaction.

Also, I had originally thought Geant4 only applied this cut at the beginning of along-step (based on the pre-step energy), but actually it looks like they apply it _again_ after slowing down, so I guess this is what we should have been doing all along.

Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: Well damn that would explain it. I meant to go back and look at that code yesterday evening... I'm going to do another search for `G4EmParameters::LowestElectronEnergy` and see if that pops up anywhere.

Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: Looks like it's used in the G4EmCalculator cuts but nowhere else. It looks like the change I've proposed is equivalent to the "post-slowing-down" application of the cut.

Comment (whokion) [reply src/celeritas/global/alongstep/AlongStep.hh:129]: I am a bit confused - the lowest energy cut is used before the actual eloss calculation [here](https://geant4.kek.jp/lxr/source/processes/electromagnetic/utils/src/G4VEnergyLossProcess.cc#L812) and [after](https://geant4.kek.jp/lxr/source/processes/electromagnetic/utils/src/G4VEnergyLossProcess.cc#L914), which may not be same or equivalent to the proposed change.

**Comment by sethrj** [comment]
@whokion I've updated the energy loss code so that we're almost doing the same thing as Geant4: we apply the electron tracking cut *before* calculating eloss as well as *after*, which this PR fixes. The exception is that we *don't* apply the tracking cut at boundary crossings: this is avoid artificially stopping positrons at geometry interfaces.

See https://github.com/celeritas-project/celeritas/blob/c9909149778dbb97b3369b240d885a71cabeac84/src/celeritas/global/alongstep/detail/MeanELoss.hh

File: src/celeritas/global/alongstep/AlongStep.hh [hunk 4]
```diff
@@ -108,17 +109,56 @@ inline CELER_FUNCTION void along_step(MH&& msc,
...
-    CELER_ASSERT(local.step_limit.action);
+    if (eloss.is_applicable(track))
+    {
+        using Energy = ParticleTrackView::Energy;
+
+        bool apply_cut = (local.step_limit.action != track.boundary_action());
+        Energy deposited
+            = eloss.calc_eloss(track, local.step_limit.step, apply_cut);
+        CELER_ASSERT(deposited <= particle.energy());
+        CELER_ASSERT(apply_cut || deposited < particle.energy());
```
Comment (whokion) [inline src/celeritas/global/alongstep/AlongStep.hh:120]: I see a case that this assert fails in demo-looop (after relaxing the max_nsteps in FieldDriver).

**Review by whokion — COMMENTED** [review]
One minor comment: Why do we use a separate MeanELoss instead of using FluctELoss with a fluctuation (off for this test) (with changing the class name FluctEloss to more generic something like EnergyLossCalculator)?

**Review by whokion — APPROVED** [review]

**Comment by sethrj** [comment]
@whokion The different eloss helper classes are because they're a template parameter to the along-step kernel: so if there are regions/energies/performance comparisons we want to make with fluctuations off, we can use that optimization. I know the AdePT team have had poor performance with Eloss on GPU so for performance comparisons we can drop it too.

File: src/celeritas/global/alongstep/AlongStep.hh (hunk 4 above)
Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:120]: I don't understand what you mean: the demo-loop unit test with adjusted parameters?

Comment (sethrj) [reply src/celeritas/global/alongstep/AlongStep.hh:120]: (This was in cms2018 through a standalone GPU demo.)

**Comment by sethrj** [comment]
OK, the current failure and the one @whokion reported may be because on master, running with `AlongStepGeneralLinearAction` bumps particles off of geometry boundaries if they slow down to zero *even if energy fluctuation is disabled*. I'm still digging in a bit...
